mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,804 @@
+      // We need to walk the function again since the structure of loop nests
+      // within the funcOp body has changed.
+      break;
+    }
+    loopNest.clear();
+    forTree.clear();
+    forOperations.clear();
+  }
+  return;
+}
+
+/// Scans the loop nest rooted at `rootForOp` and collects all affine.load and
+/// affine.store ops. Fills `loadAndStoreOps` with all such ops.
+static void getAllLoadStores(AffineForOp rootForOp,
+                             SmallVector<Operation *, 8> &loadAndStoreOps) {
+  rootForOp.getOperation()->walk([&](Operation *op) {
+    if (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op)) {
+      loadAndStoreOps.push_back(op);
+    }
+  });
+}
+
+/// Fills `elementsSize` with size of element types of respective memrefs
+/// accessed by respective ops in `loadAndStoreOps`. These will be used to 
+/// check if two accesses are within a cache_line_size/element_size distance
+/// apart for a useful locality.
+static void getElementsSize(SmallVector<Operation *, 8> &loadAndStoreOps,
+                            DenseMap<Operation *, unsigned> &elementsSize) {
+  MemRefType memrefType;
+  for (Operation *op : loadAndStoreOps) {
+    if (isa<AffineLoadOp>(op)) {
+      AffineLoadOp loadOp = cast<AffineLoadOp>(*op);
+      memrefType = loadOp.getMemRefType();
+    } else if (isa<AffineStoreOp>(op)) {
+      AffineStoreOp storeOp = cast<AffineStoreOp>(*op);
+      memrefType = storeOp.getMemRefType();
+    }
+    // If the memref has a static shape, obtain the element size. Otherwise
+    // consider a default value.
+    elementsSize[op] = memrefType.hasStaticShape()
+                           ? getMemRefSizeInBytes(memrefType).getValue() /
+                                 memrefType.getNumElements()
+                           : 8;
+  }
+}
+
+/// Fills `validPermutations` with all valid loop interchange permutations of the
+/// loop nest rooted at `rootForOp`.
+/// [Theorem] A permutation of the loops in a perfect nest is legal if and only
+/// if the direction matrix, after the same permutation is applied to its
+/// columns, has no ">" direction as the leftmost non-"=" direction in any row.
+static void getAllValidLoopInterchangePermutations(
+    AffineForOp &rootForOp, unsigned noLoopsInNest,
+    SmallVector<SmallVector<unsigned, 4>, 8> &validPermutations) {
+
+  SmallVector<unsigned, 4> permutation;
+  for (unsigned i = 0; i < noLoopsInNest; i++)
+    permutation.push_back(i);
+
+  validPermutations.push_back(permutation);
+  SmallVector<AffineForOp, 4> perfectLoopNest;
+  getPerfectlyNestedLoops(perfectLoopNest, rootForOp);
+  while (std::next_permutation(permutation.begin(), permutation.end()))
+    if (isValidLoopInterchangePermutation(perfectLoopNest, permutation))
+      validPermutations.push_back(permutation);
+}
+
+/// Calculates loop-carried-dependence vector for this loop nest rooted at
+/// `rootForOp`. Result is stored in  `loopCarriedDependenceVector`.
```

Comment:
Instead of the obvious `Result is stored in ...` say what the result means: `... is set to true if ...`.

Reply:
Fixed. Thanks.